File: ebilet_react_typescript/src/core/apiService.ts
import {parseTicket, type Ticket} from "../models/Ticket";
import { ApiClient } from "./apiClient";
import type {ApiResponse} from "../models/ApiResponse.ts";
import {parseUser, type User} from "../models/User.ts";
import {type City, parseCity} from "../models/City.ts";
import {type District, parseDistrict} from "../models/District.ts";
import {type AppEvent, parseEvent } from "../models/Event";

export class ApiService {
    static async register(params: {
        fullName: string;
        email: string;
        phone: string;
        password: string;
    }): Promise<ApiResponse<User>> {
        const json = await ApiClient.post<ApiResponse<any>>("auth/register.php", {
            full_name: params.fullName,
            email: params.email,
            phone: params.phone,
            password: params.password,
        });

        return {
            success: json.success,
            message: json.message,
            data: json.data ? parseUser(json.data) : null,
        };
    }

    static async login(params: {
        email: string;
        password: string;
    }): Promise<ApiResponse<User>> {
        const json = await ApiClient.post<ApiResponse<any>>("auth/login.php", {
            email: params.email,
            password: params.password,
        });

        return {
            success: json.success,
            message: json.message,
            data: json.data ? parseUser(json.data) : null,
        };
    }

    static async profile(apiToken: string): Promise<ApiResponse<User>> {
        const json = await ApiClient.post<ApiResponse<any>>("auth/profile.php", {
            api_token: apiToken,
        });

        return {
            success: json.success,
            message: json.message,
            data: json.data ? parseUser(json.data) : null,
        };
    }

    static async getCities(apiToken: string): Promise<ApiResponse<City[]>> {
        const json = await ApiClient.post<ApiResponse<any[]>>(
            "locations/cities_list.php",
            {
                api_token: apiToken,
            }
        );

        return {
            success: json.success,
            message: json.message,
            data: Array.isArray(json.data) ? json.data.map(parseCity) : [],
        };
    }

    static async getDistrictsByCity(params: {
        apiToken: string;
        cityId: number;
    }): Promise<ApiResponse<District[]>> {
        const json = await ApiClient.post<ApiResponse<any[]>>(
            "locations/districts_by_city.php",
            {
                api_token: params.apiToken,
                city_id: String(params.cityId),
            }
        );

        return {
            success: json.success,
            message: json.message,
            data: Array.isArray(json.data) ? json.data.map(parseDistrict) : [],
        };
    }


    static async getEventsByLocation(params: {
        apiToken: string;
        cityId: number;
        districtId: number;
    }): Promise<ApiResponse<AppEvent[]>> {
        const json = await ApiClient.post<ApiResponse<any[]>>(
            "events/events_by_location.php",
            {
                api_token: params.apiToken,
                city_id: String(params.cityId),
                district_id: String(params.districtId),
            }
        );

        return {
            success: json.success,
            message: json.message,
            data: Array.isArray(json.data) ? json.data.map(parseEvent) : [],
        };
    }

    static async getEventDetail(params: {
        apiToken: string;
        eventId: number;
    }): Promise<ApiResponse<AppEvent>> {
        const json = await ApiClient.post<ApiResponse<any>>(
            "events/event_detail.php",
            {
                api_token: params.apiToken,
                event_id: String(params.eventId),
            }
        );

        return {
            success: json.success,
            message: json.message,
            data: json.data ? parseEvent(json.data) : null,
        };
    }

    static async buyTicket(params: {
        apiToken: string;
        eventId: number;
    }): Promise<ApiResponse<Ticket>> {
        const json = await ApiClient.post<ApiResponse<any>>(
            "tickets/ticket_buy.php",
            {
                api_token: params.apiToken,
                event_id: String(params.eventId),
            }
        );

        return {
            success: json.success,
            message: json.message,
            data: json.data ? parseTicket(json.data) : null,
        };
    }

    static async getMyTickets(apiToken: string): Promise<ApiResponse<Ticket[]>> {
        const json = await ApiClient.post<ApiResponse<any[]>>(
            "tickets/my_tickets.php",
            {
                api_token: apiToken,
            }
        );

        return {
            success: json.success,
            message: json.message,
            data: Array.isArray(json.data) ? json.data.map(parseTicket) : [],
        };
    }

    static async getTicketDetail(params: {
        apiToken: string;
        ticketId: number;
    }): Promise<ApiResponse<Ticket>> {
        const json = await ApiClient.post<ApiResponse<any>>(
            "tickets/ticket_detail.php",
            {
                api_token: params.apiToken,
                ticket_id: String(params.ticketId),
            }
        );

        return {
            success: json.success,
            message: json.message,
            data: json.data ? parseTicket(json.data) : null,
        };
    }
}

File: EBiletKotlin/app/src/main/java/com/alperensarac/ebiletkotlin/data/api/ApiService.kt
package com.alperensarac.ebiletkotlin.data.api


import com.alperensarac.ebiletkotlin.data.model.ApiResponse
import com.alperensarac.ebiletkotlin.data.model.City
import com.alperensarac.ebiletkotlin.data.model.District
import com.alperensarac.ebiletkotlin.data.model.Event
import com.alperensarac.ebiletkotlin.data.model.Ticket
import com.alperensarac.ebiletkotlin.data.model.User
import com.alperensarac.ebiletkotlin.data.model.Venue
import retrofit2.Call
import retrofit2.http.Field
import retrofit2.http.FormUrlEncoded
import retrofit2.http.POST

/*
    ApiService

    Retrofit burada hangi PHP dosyasına hangi parametrelerle
    istek atacağını öğrenir.

    Örnek:

    @POST("auth/login.php")
    fun login(...)

    Bu şu URL'ye gider:
    BASE_URL + "auth/login.php"

    Yani:
    http://10.0.2.2/event_ticket_api/auth/login.php
*/
interface ApiService {

    /*
        Kullanıcı kayıt API'si.

        PHP:
        auth/register.php

        Beklenen POST:
        - full_name
        - email
        - phone
        - password
    */
    @FormUrlEncoded
    @POST("auth/register.php")
    fun register(
        @Field("full_name") fullName: String,
        @Field("email") email: String,
        @Field("phone") phone: String,
        @Field("password") password: String
    ): Call<ApiResponse<User>>

    /*
        Kullanıcı giriş API'si.

        PHP:
        auth/login.php

        Beklenen POST:
        - email
        - password
    */
    @FormUrlEncoded
    @POST("auth/login.php")
    fun login(
        @Field("email") email: String,
        @Field("password") password: String
    ): Call<ApiResponse<User>>

    /*
        Profil bilgisi.

        PHP:
        auth/profile.php

        Beklenen POST:
        - api_token
    */
    @FormUrlEncoded
    @POST("auth/profile.php")
    fun profile(
        @Field("api_token") apiToken: String
    ): Call<ApiResponse<User>>

    /*
        Şehirleri listeleme.

        PHP:
        locations/cities_list.php
    */
    @FormUrlEncoded
    @POST("locations/cities_list.php")
    fun getCities(
        @Field("api_token") apiToken: String
    ): Call<ApiResponse<List<City>>>

    /*
        Şehre göre ilçeleri listeleme.

        PHP:
        locations/districts_by_city.php
    */
    @FormUrlEncoded
    @POST("locations/districts_by_city.php")
    fun getDistrictsByCity(
        @Field("api_token") apiToken: String,
        @Field("city_id") cityId: Int
    ): Call<ApiResponse<List<District>>>

    /*
        İlçeye göre mekanları listeleme.

        PHP:
        locations/venues_by_district.php
    */
    @FormUrlEncoded
    @POST("locations/venues_by_district.php")
    fun getVenuesByDistrict(
        @Field("api_token") apiToken: String,
        @Field("city_id") cityId: Int,
        @Field("district_id") districtId: Int
    ): Call<ApiResponse<List<Venue>>>

    /*
        Şehir + ilçe seçimine göre etkinlikleri listeleme.

        PHP:
        events/events_by_location.php

        venue_id opsiyonel demiştik ama Retrofit'te sade tutmak için
        şimdilik göndermiyoruz.
    */
    @FormUrlEncoded
    @POST("events/events_by_location.php")
    fun getEventsByLocation(
        @Field("api_token") apiToken: String,
        @Field("city_id") cityId: Int,
        @Field("district_id") districtId: Int
    ): Call<ApiResponse<List<Event>>>

    /*
        Etkinlik detayı.

        PHP:
        events/event_detail.php
    */
    @FormUrlEncoded
    @POST("events/event_detail.php")
    fun getEventDetail(
        @Field("api_token") apiToken: String,
        @Field("event_id") eventId: Int
    ): Call<ApiResponse<Event>>

    /*
        Bilet satın alma.

        PHP:
        tickets/ticket_buy.php
    */
    @FormUrlEncoded
    @POST("tickets/ticket_buy.php")
    fun buyTicket(
        @Field("api_token") apiToken: String,
        @Field("event_id") eventId: Int
    ): Call<ApiResponse<Ticket>>

    /*
        Biletlerim.

        PHP:
        tickets/my_tickets.php
    */
    @FormUrlEncoded
    @POST("tickets/my_tickets.php")
    fun getMyTickets(
        @Field("api_token") apiToken: String
    ): Call<ApiResponse<List<Ticket>>>

    /*
        Bilet detay.

        PHP:
        tickets/ticket_detail.php
    */
    @FormUrlEncoded
    @POST("tickets/ticket_detail.php")
    fun getTicketDetail(
        @Field("api_token") apiToken: String,
        @Field("ticket_id") ticketId: Int
    ): Call<ApiResponse<Ticket>>

    /*
        QR bilet kontrol.

        PHP:
        check/ticket_check.php

        Bu API sadece staff veya admin tokenı ile çalışır.
    */
    @FormUrlEncoded
    @POST("check/ticket_check.php")
    fun checkTicket(
        @Field("api_token") apiToken: String,
        @Field("ticket_code") ticketCode: String
    ): Call<ApiResponse<Ticket>>
}